### PR TITLE
[FLINK-34917][table] Introduce comment for CatalogStore & Support enhanced `CREATE CATALOG` syntax

### PR DESCRIPTION
**Pull request by liyubin117**
## What is the purpose of the change

Introduce comment for CatalogStore and Support enhanced `CREATE CATALOG` syntax

## Brief change log

* add comment instance in `CatalogDescriptor` and expose setter/getter function
* CREATE CATALOG [IF NOT EXISTS] catalog_name [COMMENT 'comment_value'] [WITH (property_name=property_value, ...)]

## Verifying this change

This change added tests and can be verified as follows:

flink-table/flink-sql-client/src/test/resources/sql/catalog_database.q
flink-table/flink-sql-gateway/src/test/resources/sql/catalog_database.q
flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest#testCatalogStore
flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest#testAlterCatalog

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? yes

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java [hunk 1]
```diff
@@ -53,11 +63,23 @@ public Map<String, String> getProperties() {
         return properties;
     }
 
+    public @Nullable String getComment() {
+        return comment;
+    }
```
Comment (LadyForest) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java:68]: Remove this getter if there is no use.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlCreateCatalogConverter.java [hunk 2]
```diff
@@ -41,6 +43,10 @@ public Operation convertSqlNode(SqlCreateCatalog node, ConvertContext context) {
                                         ((SqlTableOption) p).getKeyString(),
                                         ((SqlTableOption) p).getValueString()));
 
-        return new CreateCatalogOperation(node.catalogName(), properties);
+        return new CreateCatalogOperation(
+                node.catalogName(),
+                properties,
+                getCatalogComment(node.getComment()),
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlCreateCatalogConverter.java:49]:
```suggestion
                node.getComment().map(c -> c.getValueAs(String.class)).orElse(null),
```

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 3]
```diff
@@ -85,6 +87,13 @@ public static List<TableChange> buildModifyColumnChange(
         }
     }
 
+    public static @Nullable String getCatalogComment(Optional<SqlNode> catalogComment) {
+        return catalogComment
+                .map(SqlCharStringLiteral.class::cast)
+                .map(c -> c.getValueAs(NlsString.class).getValue())
+                .orElse(null);
+    }
+
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:96]: it can be inlined, so remove this

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 4]
```diff
@@ -295,31 +295,42 @@ public DataTypeFactory getDataTypeFactory() {
...
             throws CatalogException {
         checkArgument(
                 !StringUtils.isNullOrWhitespaceOnly(catalogName),
                 "Catalog name cannot be null or empty.");
         checkNotNull(catalogDescriptor, "Catalog descriptor cannot be null");
 
         if (catalogStoreHolder.catalogStore().contains(catalogName)) {
-            throw new CatalogException(
-                    format("Catalog %s already exists in catalog store.", catalogName));
+            if (!ignoreIfExists) {
```
Comment (LadyForest) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:312]: The logic is a little bit cumbersome, what about

```java
public void createCatalog(
        String catalogName, CatalogDescriptor catalogDescriptor, boolean ignoreIfExists)
        throws CatalogException {
    checkArgument(
            !StringUtils.isNullOrWhitespaceOnly(catalogName),
            "Catalog name cannot be null or empty.");
    checkNotNull(catalogDescriptor, "Catalog descriptor cannot be null");

    boolean catalogExistsInStore = catalogStoreHolder.catalogStore().contains(catalogName);
    boolean catalogExistsInMemory = catalogs.containsKey(catalogName);

    if (catalogExistsInStore || catalogExistsInMemory) {
        if (!ignoreIfExists) {
            throw new CatalogException(
                    format("Catalog %s already exists.", catalogName));
        }
    } else {
        // Store the catalog in the catalog store
        catalogStoreHolder.catalogStore().storeCatalog(catalogName, catalogDescriptor);
        
        // Initialize and store the catalog in memory
        Catalog catalog = initCatalog(catalogName, catalogDescriptor);
        catalog.open();
        catalogs.put(catalogName, catalog);
    }
}

```

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java [hunk 5]
```diff
@@ -367,14 +368,28 @@ void testCatalogStore() throws Exception {
         catalogManager.createCatalog("cat1", CatalogDescriptor.of("cat1", configuration));
         catalogManager.createCatalog("cat2", CatalogDescriptor.of("cat2", configuration));
         catalogManager.createCatalog("cat3", CatalogDescriptor.of("cat3", configuration));
+        catalogManager.createCatalog(
```
Comment (LadyForest) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java:371]: Please add the exception test as well.
```java
    assertThatThrownBy(
                        () ->
                                catalogManager.createCatalog(
                                        "cat_comment",
                                        CatalogDescriptor.of(
                                                "cat_comment",
                                                configuration.clone(),
                                                "second comment for catalog"),
                                        false))
                .isInstanceOf(CatalogException.class)
                .hasMessage("Catalog cat_comment already exists in catalog store.");
```

File: flink-core/src/test/java/org/apache/flink/core/io/LocatableSplitAssignerTest.java [hunk 6]
```diff
@@ -428,7 +428,7 @@ void testConcurrentSplitAssignmentForMultipleHosts() throws InterruptedException
         assertThat(ia.getNextInputSplit("testhost", 0)).isNull();
 
         // at least one fraction of hosts needs be local, no matter how bad the thread races
-        assertThat(ia.getNumberOfRemoteAssignments())
+        assertThat(ia.getNumberOfLocalAssignments())
```
Comment (LadyForest) [inline flink-core/src/test/java/org/apache/flink/core/io/LocatableSplitAssignerTest.java:431]: Do not include commits that don't belong to this PR.

**Review by LadyForest — COMMENTED** [review]
Thanks for your contribution. I left some comments, PTAL

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateCatalog.java [hunk 7]
```diff
@@ -45,11 +48,18 @@ public class SqlCreateCatalog extends SqlCreate {
 
     private final SqlNodeList propertyList;
 
+    private final @Nullable SqlNode comment;
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateCatalog.java]: StringLiteral() returns SqlNode, so we should keep it as `SqlNode`
baf72e629b5f98822abd464ce74da00

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogDescriptor.java [hunk 8]
```diff
@@ -48,18 +55,34 @@ public Configuration getConfiguration() {
...
     }
 
-    private CatalogDescriptor(String catalogName, Configuration configuration) {
+    public Optional<String> getComment() {
+        return Optional.ofNullable(comment);
+    }
+
+    public CatalogDescriptor setComment(String comment) {
+        return new CatalogDescriptor(catalogName, configuration, comment);
+    }
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogDescriptor.java]: done

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 9]
```diff
@@ -134,6 +134,28 @@ void testCreateCatalog() {
...
+                        + " WITH (\n"
+                        + "  'key1'='value1',\n"
+                        + "  'key2'='value2'\n"
+                        + " )\n")
+                .ok(
+                        "CREATE CATALOG IF NOT EXISTS `C1`\n"
+                                + "COMMENT 'hello' WITH (\n"
+                                + "  'key1' = 'value1',\n"
+                                + "  'key2' = 'value2'\n"
+                                + ")");
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java]: done

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java [hunk 10]
```diff
@@ -39,10 +41,18 @@
 public class CreateCatalogOperation implements CreateOperation {
     private final String catalogName;
     private final Map<String, String> properties;
+    private final @Nullable String comment;
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java]: done

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java (hunk 1 above)
Comment (liyubin117) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateCatalogOperation.java]: done

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlCreateCatalogConverter.java (hunk 2 above)
Comment (liyubin117) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlCreateCatalogConverter.java]: done

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java (hunk 3 above)
Comment (liyubin117) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java]: done

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java (hunk 4 above)
Comment (liyubin117) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:315]: good idea

File: flink-core/src/test/java/org/apache/flink/core/io/LocatableSplitAssignerTest.java (hunk 6 above)
Comment (liyubin117) [reply flink-core/src/test/java/org/apache/flink/core/io/LocatableSplitAssignerTest.java]: done

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java (hunk 5 above)
Comment (liyubin117) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java:371]: done

**Review by LadyForest — APPROVED** [review]
LGTM
